tools/webdriver/webdriver/bidi/client.py

```diff
@@ -0,0 +1,264 @@
+
+        # Modules
+        self.session = Session(self)
+
+        if loop is None:
+            loop = asyncio.get_running_loop()
+        self.loop = loop
+
+    @staticmethod
+    def set_capability(caps):
+        if caps:
+            caps.setdefault("alwaysMatch", {}).update({"webSocketUrl": True})
+        else:
+            caps = {"alwaysMatch": {"webSocketUrl": True}}
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.end()
+
+    async def start(self):
+        """Start a new WebDriver Bidirectional session
+        with websocket connected.
+        """
+        if urlparse(self.websocket_url).path == "/":
+            if self.session_id is None:
+                self.websocket_url = urljoin(self.websocket_url, "session")
+            else:
+                self.websocket_url = urljoin(self.websocket_url, "session/%s" % self.session_id)
+        self.transport = Transport(self.websocket_url, self.on_message, loop=self.loop)
+
+        if self.session_id is None:
+            await self.session.new(self.capabilities)
+
+        await self.transport.start()
+
+    async def send_command(self, method, params):
+        # this isn't threadsafe
+        self.command_id += 1
+        command_id = self.command_id
+
+        body = {
+            "id": command_id,
+            "method": method,
+            "params": params
+        }
+        self.pending_commands[command_id] = self.loop.create_future()
+        await self.transport.send(body)
+
+        return self.pending_commands[command_id]
+
+    async def on_message(self, data):
+        if "id" in data:
+            # This is a command response or error
+            future = self.pending_commands.get(data["id"])
+            if future is None:
+                raise ValueError("No pending command with id %s" % data["id"])
+            if "result" in data:
+                future.set_result(data["result"])
+            elif "error" in data and "message" in data:
+                future.set_exception(BidiException(data["error"],
+                                                   data["message"],
+                                                   data.get("stacktrace")))
+            else:
+                raise ValueError("Unexpected message: %r" % data)
+        elif "method" in data and "params" in data:
+            # This is an event
+            method = data["method"]
+            listeners = self.event_listeners.get(method)
+            if not listeners:
+                listeners = self.event_listeners[None]
+            for listener in listeners:
+                await listener(data["params"])
+        else:
+            raise ValueError("Unexpected message: %r" % data)
+
+    async def end(self):
+        """Close websocket connection first before closing session.
+        """
+        await self.transport.end()
+
+    def add_event_listener(self, name, fn):
```

Comment:
Could you add a test demonstrating event listener?

Reply:
Not easily since I don't have something to test it against yet. Does it work with the Chrome prototype and how can I run that?

Reply:
Sorry for the late reply. Chrome does not support it yet.